### PR TITLE
[ARMISelDAGToDAG] use MO_FrameIndex to represent FrameIndex rather than MO_Register for INLINEASM

### PR DESCRIPTION
**Pull request by nickdesaulniers**
Many other targets use stack slot indexes to represent memory locations.
This patch normalizes ARM to also use these kind of MachineOperands so
that register selection can find which stack slot is necessary to fold
memory operands when "rm" is used.

A nice side effect of this change is that we're less likely to exhaust
registers for inline asm with many clobbers.

Link: https://github.com/llvm/llvm-project/issues/20571#issuecomment-1771336576

### CODE REVIEW

File: llvm/lib/Target/ARM/ARMBaseInstrInfo.cpp [hunk 1]
```diff
@@ -2649,9 +2649,9 @@ bool llvm::rewriteARMFrameIndex(MachineInstr &MI, unsigned FrameRegIdx,
   unsigned AddrMode = (Desc.TSFlags & ARMII::AddrModeMask);
   bool isSub = false;
 
-  // Memory operands in inline assembly always use AddrMode2.
+  // Memory operands in inline assembly always use AddrMode_i12.
   if (Opcode == ARM::INLINEASM || Opcode == ARM::INLINEASM_BR)
-    AddrMode = ARMII::AddrMode2;
+    AddrMode = ARMII::AddrMode_i12;
```
Comment (nickdesaulniers) [inline llvm/lib/Target/ARM/ARMBaseInstrInfo.cpp:2654]: this needs the same `InlineAsm::Flag` check as ARMAsmPrinter

**Comment by efriedma-quic** [comment]
I suspect this will trigger assembler errors in some cases.  ARM has a variety of addressing modes, and most of them aren't compatible with SelectAddrModeImm12.  Particularly for the constraints that are not "m"; "Uv" is explicitly documented to be an operand for a "vldr", which only has a 10 bit offset.

And it looks like the handling for the case where the the stack frame is actually larger than 4096 bytes is missing?

**Comment by nickdesaulniers** [comment]
> Particularly for the constraints that are not "m"; "Uv" is explicitly documented to be an operand for a "vldr", which only has a 10 bit offset.

Yeah, I was waffling about splitting the cases in `ARMDAGToDAGISel::SelectInlineAsmMemoryOperand` to just do this change for the non-`Ux` constraints.  I could change it back.

> And it looks like the handling for the case where the the stack frame is actually larger than 4096 bytes is missing?

Is that possible to know (the current stack depth) during SelectionDAGISel.  Or should that be handled elsewhere later? GCC just errors once the stack gets too large.

**Comment by nickdesaulniers** [comment]
I'll be reusing pieces of this change in a follow up PR, but for now, I won't change over arm in general to do this.

**Comment by efriedma-quic** [comment]
> Is that possible to know (the current stack depth) during SelectionDAGISel? Or should that be handled elsewhere later?

You can sort of tell some things; for example, I think we track whether the function has dynamic allocation.  But the stack layout isn't finalized until FrameLowering.

**Pull request closed without merging**